etf/encode.go

```diff
@@ -529,10 +543,11 @@ func Encode(term Term, b *lib.Buffer,
 
 		case []byte:
 			lenBinary := len(t)
-			buf := b.Extend(1 + 4 + lenBinary)
-			buf[0] = ettBinary
+			buf := b.Extend(1 + 4 + 1 + lenBinary)
+			buf[0] = ettBitBinary
```

Comment:
I see your idea to use ettBinary as a transport for the string and ettBitBinary for the real binary data but it makes Ergo-Ergo interaction a bit harder for the case of string usage or for the case if I send real binary (not a string) from the Erlang side.

Reply:
This commit was based on your previous input that this library's priority is Ergo <-> Ergo.

If we start considering the case for Erlang, we should also consider the case for Elixir. Elixir strings are binaries.

Do also note that the current ergo implementation can only send ASCII strings to Erlang and there are no safety checks to ensure that the user passes only ASCII goStrings.
Rather than have golang side waste CPU cycles to check if a string contains utf8 or not, it's why `etf.String` was added to support sending legacy style ASCII-only strings to Erlang.

| Ergo ->  | transport -> | Elixir | transport -> | Ergo |
| -------- | ------------ | ----- | ------------ | ------- |
| []byte | ettBitBinary | <<.....>> | ettBitBinary | []byte |
| string (ascii) | ettBinary | "...." | ettBinary | string |
| string (utf8) | ettBinary | "...." | ettBinary | string |
| etf.String | ettString | '....' or [ <uint> , <uint>, <uint> ] | ettString | etf.String |

Note single quotes in Elixir produce charlists (and only support ascii characters), unlike double quotes .
Charlists are defined as a linked list of positive integers that can use [ h | tail ] pattern matching. (Charlist is not a concrete type in elixir).

ettString automatically becomes a charlist in Elixir (and are displayed as string with single quotes in Elixir shell).
List of positive integers (charlist) are also displayed as string with quotes in Erlang shell: https://erlang.org/doc/apps/stdlib/unicode_usage.html#heuristic-string-detection

| Ergo ->  | transport -> | Ergo |
| -------- | ------------ | ----- |
| string (utf8) | ettBinary | string |
| byte[] | ettBitBinary | []byte |

Reply:
not sure about Elixir <<...>>> -> ettBitBinary. May I ask you to show the same output 
```
16> term_to_binary(<<1,2,3>>).
<<131,109,0,0,0,3,1,2,3>>
```
but in Elixir shell? (I'm not familiar with it)

Reply:
just found
```
iex(1)> :erlang.term_to_binary(<<1,2,3>>);
<<131, 109, 0, 0, 0, 3, 1, 2, 3>>
```
as you may notice it was encoded as ettBinary (109)

Reply:
Elixir:
```elixir
Interactive Elixir (1.12.2) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> <<195,165,195,164,195,182>> 
"åäö"
iex(2)> :erlang.term_to_binary("åäö")
<<131, 109, 0, 0, 0, 6, 195, 165, 195, 164, 195, 182>>
iex(3)> :erlang.term_to_binary("123")
<<131, 109, 0, 0, 0, 3, 49, 50, 51>>
iex(4)> :erlang.term_to_binary("日本")
<<131, 109, 0, 0, 0, 6, 230, 151, 165, 230, 156, 172>>
```

Erlang:
```erlang
Eshell V10.7.2.12  (abort with ^G)
1> <<195,165,195,164,195,182>>.
<<"åäö"/utf8>>
2> term_to_binary(<<"åäö"/utf8>>).
<<131,109,0,0,0,6,195,165,195,164,195,182>>
3> term_to_binary("åäö").
<<131,107,0,3,229,228,246>>
4> term_to_binary("123"). 
<<131,107,0,3,49,50,51>>
5> term_to_binary("日本").
<<131,108,0,0,0,2,98,0,0,101,229,98,0,0,103,44,106>>
```

I do understand your point, but it's not a coincidence that we can Use Strings as Byte Slices in golang: https://go101.org/article/string.html#use-string-as-byte-slice


Since `string` is just an immutable `[]byte` according to Rob Pike...

So i think the question would be, should decoded binary values be immutable or mutable?
Does immutability in this case help prevent a class programming bugs?

This topic is still a big contention even within the Golang Issue Tracker:
See "Strengths of This Proposal" from: 
- https://github.com/golang/go/issues/22876
- https://github.com/golang/go/issues/37303
- https://github.com/golang/go/issues/32245
- https://github.com/golang/go/issues/27975

Which is why I would approach it from language design: what is a string in Elixir, what is a string in Erlang, and what is a String in Golang?

And i find that the common denominator is that strings are just a sequence of bytes in all three languages.

Reply:
In other words, 107 is just an optimisation on 108.

Reply:
Shall we just make this configurable?

Reply:
I appreciate your effort to make this project better, but seemingly this approach differs from the way this project goes.